File: Blind 75/06 - Linked List/19-remove-nth-node-from-end-of-list.py
'''
Given the head of a linked list, remove the nth node from the end of the list and return its head.
'''
from time import time
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        node = ListNode(0, head)
        p1, p2 = node, head
        # Adjust gap between pointers
        for i in range(n):
            p2 = p2.next

        while p2:
            p1 = p1.next
            p2 = p2.next
        p1.next = p1.next.next

        return node.next

    def reference(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        left = dummy
        right = head

        while n > 0:
            right = right.next
            n -= 1

        while right:
            left = left.next
            right = right.next

        # Delete
        left.next = left.next.next
        return dummy.next

    def quantify(self, test_cases, runs=100000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.removeNthFromEnd(case[0], case[1]))
                else:
                    self.removeNthFromEnd(case[0], case[1])
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case[0], case[1]))
                else:
                    self.reference(case[0], case[1])
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        (ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5))))), 2),
        (ListNode(1), 1),
        (ListNode(1, ListNode(2)), 1),
    ]
    test.quantify(test_cases)

File: Blind 75/06 - Linked List/JavaScript/19-remove-nth-node-from-end-of-list.js
const { performance } = require('perf_hooks');

class ListNode {
  constructor(val, next) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

class Solution {
  method(head, n) {
    let node = new ListNode();
    node.next = head;
    let [p1, p2] = [node, head];
    for (let i = 0; i < n; i++) {
      p2 = p2.next;
    }
    while (p2) {
      p1 = p1.next;
      p2 = p2.next;
    }
    p1.next = p1.next.next;

    return node.next;
  }

  reference(head, n) {
    const length = this.getNthFromEnd(head, n); /* Time O(N) */

    const isHead = length < 0;
    if (isHead) return head.next;

    const curr = this.moveNode(head, length); /* Time O(N) */

    curr.next = curr.next.next;

    return head;
  }

  getNthFromEnd = (curr, n, length = 0) => {
    while (curr) {
      /* Time O(N) */
      curr = curr.next;
      length++;
    }

    return length - n - 1;
  };

  moveNode = (curr, length) => {
    while (length) {
      /* Time O(N) */
      curr = curr.next;
      length--;
    }

    return curr;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        // Create deep copy
        const copy = JSON.parse(JSON.stringify(input[0]));
        if (i === 0) console.log(this.method(copy, input[1]));
        else this.method(copy, input[1]);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        // Create deep copy
        const copy = JSON.parse(JSON.stringify(input[0]));
        if (i === 0) console.log(this.reference(copy, input[1]));
        else this.reference(copy, input[1]);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [
    new ListNode(
      1,
      new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5))))
    ),
    2,
  ],
  [new ListNode(1), 1],
  [new ListNode(1, new ListNode(2)), 1],
];
test.quantify(testCases);
